src/bsoncxx/stdx/view_or_value.hpp

```diff
@@ -0,0 +1,70 @@
+// Copyright 2015 MongoDB Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <bsoncxx/config/prelude.hpp>
+
+namespace bsoncxx {
+BSONCXX_INLINE_NAMESPACE_BEGIN
+namespace stdx {
+
+///
+/// Class representing a view-or-value variant type. Class View must be constructible
+/// from an instance of class Value, and class Value must be default constructible.
+///
+template <typename View, typename Value>
+class BSONCXX_API view_or_value {
+   public:
+    ///
+    /// Construct a view_or_value from a View. When constructed with a View,
+    /// this object is non-owning. The Value underneath the given View must outlive this object.
+    ///
+    /// @param view
+    ///   A non-owning View.
+    ///
+    BSONCXX_INLINE view_or_value(View view) : _view(view) {
+    }
+
+    ///
+    /// Constructs a view_or_value from a Value type. This object owns the passed-in Value.
+    ///
+    /// @param value
+    ///   A Value type.
+    ///
+    BSONCXX_INLINE view_or_value(Value value) : _value(std::move(value)), _view(_value) {
+    }
+
```

Comment:
You'll need the full raft of life cycle methods because the values can be copyable.  And after a copy you'll end up with a view that points at the old value.